### PR TITLE
add command decoding benchmark

### PR DESCRIPTION
**Pull request by monkey92t**
Can add more....

```go
BenchmarkDecode/RespError-Client-2                 	 1000000	      1181 ns/op	     228 B/op	       7 allocs/op
BenchmarkDecode/RespStatus-Client-2                	 1000000	      1185 ns/op	     248 B/op	       6 allocs/op
BenchmarkDecode/RespInt-Client-2                   	 1000000	      1078 ns/op	     180 B/op	       5 allocs/op
BenchmarkDecode/RespString-Client-2                	 1000000	      1123 ns/op	     208 B/op	       6 allocs/op
BenchmarkDecode/RespArray-Client-2                 	  784459	      1525 ns/op	     336 B/op	      12 allocs/op
BenchmarkDecode/RespPipeline-Client-2              	  471608	      2591 ns/op	     800 B/op	      22 allocs/op
BenchmarkDecode/RespTxPipeline-Client-2            	  338113	      3528 ns/op	    1176 B/op	      35 allocs/op
BenchmarkDecode/DynamicGoroutine-Client-pool=5-2   	  444393	      2793 ns/op	     208 B/op	       6 allocs/op
BenchmarkDecode/DynamicGoroutine-Client-pool=20-2  	  446067	      2783 ns/op	     208 B/op	       5 allocs/op
BenchmarkDecode/DynamicGoroutine-Client-pool=50-2  	  446502	      2787 ns/op	     207 B/op	       5 allocs/op
BenchmarkDecode/DynamicGoroutine-Client-pool=100-2 	  446889	      2781 ns/op	     207 B/op	       5 allocs/op
BenchmarkDecode/StaticGoroutine-Client-pool=5-2    	 1000000	      1130 ns/op	     207 B/op	       5 allocs/op
BenchmarkDecode/StaticGoroutine-Client-pool=20-2   	 1000000	      1087 ns/op	     208 B/op	       5 allocs/op
BenchmarkDecode/StaticGoroutine-Client-pool=50-2   	 1000000	      1095 ns/op	     207 B/op	       5 allocs/op
BenchmarkDecode/StaticGoroutine-Client-pool=100-2  	 1000000	      1099 ns/op	     207 B/op	       5 allocs/op
BenchmarkDecode/RespError-Cluster-2                	  805978	      1553 ns/op	     228 B/op	       7 allocs/op
BenchmarkDecode/RespStatus-Cluster-2               	  858656	      1460 ns/op	     248 B/op	       6 allocs/op
BenchmarkDecode/RespInt-Cluster-2                  	  926455	      1353 ns/op	     180 B/op	       5 allocs/op
BenchmarkDecode/RespString-Cluster-2               	  900232	      1390 ns/op	     208 B/op	       6 allocs/op
BenchmarkDecode/RespArray-Cluster-2                	  655177	      1886 ns/op	     336 B/op	      12 allocs/op
BenchmarkDecode/RespPipeline-Cluster-2             	  117696	     10202 ns/op	    1344 B/op	      30 allocs/op
BenchmarkDecode/RespTxPipeline-Cluster-2           	   96346	     13027 ns/op	    2016 B/op	      48 allocs/op
BenchmarkDecode/DynamicGoroutine-Cluster-pool=5-2  	  429230	      2697 ns/op	     207 B/op	       5 allocs/op
BenchmarkDecode/DynamicGoroutine-Cluster-pool=20-2 	  465974	      2673 ns/op	     207 B/op	       5 allocs/op
BenchmarkDecode/DynamicGoroutine-Cluster-pool=50-2 	  464558	      2725 ns/op	     207 B/op	       5 allocs/op
BenchmarkDecode/DynamicGoroutine-Cluster-pool=100-2         	  469230	      2699 ns/op	     207 B/op	       5 allocs/op
BenchmarkDecode/StaticGoroutine-Cluster-pool=5-2            	 1000000	      1264 ns/op	     207 B/op	       5 allocs/op
BenchmarkDecode/StaticGoroutine-Cluster-pool=20-2           	 1000000	      1247 ns/op	     208 B/op	       5 allocs/op
BenchmarkDecode/StaticGoroutine-Cluster-pool=50-2           	 1000000	      1213 ns/op	     208 B/op	       5 allocs/op
BenchmarkDecode/StaticGoroutine-Cluster-pool=100-2          	 1000000	      1215 ns/op	     207 B/op	       5 allocs/op
```

### CODE REVIEW

File: benchmark_decode_test.go [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+		// multi
+		"*7\r\n+multi\r\n:1\r\n*4\r\n+noscript\r\n+loading\r\n+stale\r\n+fast\r\n:0\r\n:0\r\n:0\r\n" +
+			"*2\r\n+@fast\r\n+@transaction\r\n",
+
+		// exec
+		"*7\r\n+exec\r\n:1\r\n*5\r\n+noscript\r\n+loading\r\n+stale\r\n+skip_monitor\r\n+skip_slowlog\r\n" +
+			":0\r\n:0\r\n:0\r\n*2\r\n+@slow\r\n+@transaction\r\n",
+
+		// ping
+		"*7\r\n+ping\r\n:-1\r\n*2\r\n+stale\r\n+fast\r\n:0\r\n:0\r\n:0\r\n*2\r\n+@fast\r\n+@connection\r\n",
```
Comment (vmihailenco) [inline benchmark_decode_test.go:91]: I don't understand the purpose of these commands... As I understand `conn.buff` is a response buffer, but these are outgoing commands...

Comment (monkey92t) [reply benchmark_decode_test.go:91]: The cluster uses the `command` command (cmdsInfoCache), but does not need to respond to all commands (200+), only the commands used in our test

**Comment by monkey92t** [comment]
It can provide a benchmark test that does not require redis-server, and is a performance reference when improving go-redis in the future.

File: benchmark_decode_test.go (hunk 1 above)
Comment (vmihailenco) [reply benchmark_decode_test.go:91]: Did you consider setting `client.cmdsInfoCache` instead? We could even grab `client.cmdsInfoCache` from the real cluster client so we don't need to construct it ourselves.

Comment (monkey92t) [reply benchmark_decode_test.go]: I modified to get cmdsInfoCache from the real redis-server, what I originally thought was not to rely on redis-server

File: benchmark_decode_test.go [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+		conn:   conn,
+		client: client,
+	}
+}
+
+type RespDecode struct {
+	name   string
+	conn   *BenchConn
+	client Cmdable
+}
```
Comment (vmihailenco) [inline benchmark_decode_test.go:79]: What do you think about tweaking API a bit

```go
type ClientStub struct {
    Cmdable
    conn *ConnStub
}

// Or StubResponse?
func (ClientStub) SetResponse(b []byte) {}

stub.SetResponse([]byte("something"))
stub.Ping(ctx)
```

File: benchmark_decode_test.go (hunk 1 above)
Comment (vmihailenco) [reply benchmark_decode_test.go]: Thanks, it looks much more approachable now.

File: benchmark_decode_test.go (hunk 2 above)
Comment (monkey92t) [reply benchmark_decode_test.go:112]: It looks good, I have made changes and supported large data volume testing

**Comment by monkey92t** [comment]
Support for redis.RESP3, it is difficult to maintain the current performance of RESP2. According to my test, it will cause a 5-10% performance drop, which is about an increase of 100-400ns/op(The way to use https://github.com/go-redis/redis/pull/1716).
This is because the RESP3 protocol adds an `Attribute type`, which looks like this:
```go
*3<CR><LF>
    :1<CR><LF>
    :2<CR><LF>
    |1<CR><LF>
        +ttl
        :3600
    :3<CR><LF>

> r = Redis.new
#<Redis client>

> r.mget("a","b", "c")
#<Redis reply>

> r
[1,2,3]

> r[2].attribs
{"ttl": 3600}
```

The current command->proto.Reader cannot handle it, and a processing method similar to https://github.com/go-redis/redis/pull/1716 (proto.Value) needs to be used, unless we abandon this type.

**Comment by vmihailenco** [comment]
Thanks :+1: I will do some tweaks myself too

**Pull request merged**